File: bindings/nodejs/tests/directory_extension.js
/*
 * directory_extension.js
 *
 * This source file is part of the FoundationDB open source project
 *
 * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

"use strict";

var util = require('util');
var fdb = require('../lib/fdb.js').apiVersion(parseInt(process.argv[3]));
var fdbUtil = require('../lib/fdbUtil.js');
var dirUtil = require('./directory_util.js');

var logAll = false;

var logInstructions = false;
var logOps = false;
var logDirs = false;
var logErrors = false;

var logOp = function(message, force) {
	if(logOps || logAll || force)
		console.log(message);
};

var DirectoryExtension = function() {
	this.dirList = [fdb.directory];
	this.dirIndex = 0;
	this.errorIndex = 0;
};

DirectoryExtension.prototype.processInstruction = function(inst, cb) {
	var self = this;
	var directory = this.dirList[this.dirIndex];

	var promiseCb = function(err) {
		if(err && (logErrors || logAll)) {
			console.log(err);
			//console.log(err.stack);
		}

		dirUtil.pushError(self, inst, err);
		cb();
	};

	var appendDir = function(dir) {
		if(logDirs || logAll)
			console.log(util.format('pushed at %d (op=%s)',  self.dirList.length, inst.op));

		self.dirList.push(dir);
	};

	if(logAll || logInstructions)
		console.log(inst.context.instructionIndex, inst.tokens[0].toString());

	if(inst.op === 'DIRECTORY_CREATE_SUBSPACE') {
		dirUtil.popTuples(inst)
		.then(function(path) {
			return inst.pop()
			.then(function(rawPrefix) {
				logOp(util.format('created subspace at (%s): %s',  path, fdb.buffer.printable(rawPrefix)));
				appendDir(new fdb.Subspace(path, rawPrefix));
			});
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_CREATE_LAYER') {
		inst.pop({count: 3})
		.then(function(params) {
			var index1 = params[0];
			var index2 = params[1];
			var allowManualPrefixes = params[2];

			var dir1 = self.dirList[params[0]];
			var dir2 = self.dirList[params[1]];
			if(dir1 === null || dir2 === null) {
				logOp('create directory layer: None');
				appendDir(null);
			}
			else {
				logOp(util.format('create directory layer: node_subspace (%d) = %s, content_subspace (%d) = %s, allow_manual_prefixes = %d', index1, fdb.buffer.printable(dir1.rawPrefix), index2, fdb.buffer.printable(dir2.rawPrefix), allowManualPrefixes));
				appendDir(new fdb.DirectoryLayer({ nodeSubspace: dir1,
														   contentSubspace: dir2,
														   allowManualPrefixes: allowManualPrefixes === 1 }));
			}
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_CHANGE') {
		inst.pop()
		.then(function(index) {
			if(self.dirList[index] === null)
				self.dirIndex = self.errorIndex;
			else
				self.dirIndex = index;

			if(logDirs || logAll) {
				var dir = self.dirList[self.dirIndex];
				console.log(util.format('changed directory to %d ((%s))', self.dirIndex, dir._path));
			}
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_SET_ERROR_INDEX') {
		inst.pop()
		.then(function(errorIndex) {
			self.errorIndex = errorIndex;
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_CREATE_OR_OPEN') {
		dirUtil.popTuples(inst)
		.then(function(path) {
			return inst.pop()
			.then(function(layer) {
				logOp(util.format('create_or_open (%s): layer=%s', directory._path + path, fdb.buffer.printable(layer) || ''));
				return directory.createOrOpen(inst.tr, path, {'layer': layer || undefined});
			})
			.then(function(dir) {
				appendDir(dir);
			});
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_CREATE') {
		dirUtil.popTuples(inst)
		.then(function(path) {
			return inst.pop({count: 2})
			.then(function(params) {
				logOp(util.format('create (%s): layer=%s, prefix=%s', directory._path + path, fdb.buffer.printable(params[0]) || '', fdb.buffer.printable(params[1] || '')));
				return directory.create(inst.tr, path, {'layer': params[0] || undefined, 'prefix': params[1] || undefined});
			})
			.then(function(dir) {
				appendDir(dir);
			});
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_OPEN') {
		dirUtil.popTuples(inst)
		.then(function(path) {
			return inst.pop()
			.then(function(layer) {
				logOp(util.format('open (%s): layer=%s', directory._path + path, fdb.buffer.printable(layer) || ''));
				return directory.open(inst.tr, path, {'layer': layer});
			})
			.then(function(dir) {
				appendDir(dir);
			});
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_MOVE') {
		dirUtil.popTuples(inst, 2)
		.then(function(paths) {
			logOp(util.format('move (%s) to (%s)', directory._path + paths[0], directory._path + paths[1]));
			return directory.move(inst.tr, paths[0], paths[1])
			.then(function(dir) {
				appendDir(dir);
			});
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_MOVE_TO') {
		dirUtil.popTuples(inst)
		.then(function(newAbsolutePath) {
			logOp(util.format('move (%s) to (%s)', directory._path, newAbsolutePath));
			return directory.moveTo(inst.tr, newAbsolutePath)
			.then(function(dir) {
				appendDir(dir);
			});
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_REMOVE') {
		inst.pop()
		.then(function(count) {
			return dirUtil.popTuples(inst, count)
			.then(function(path) {
				logOp(util.format('remove (%s)', directory._path + (path ? path : '')));
				return directory.remove(inst.tr, path);
			});
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_REMOVE_IF_EXISTS') {
		inst.pop()
		.then(function(count) {
			return dirUtil.popTuples(inst, count)
			.then(function(path) {
				logOp(util.format('remove_if_exists (%s)', directory._path + (path ? path : '')));
				return directory.removeIfExists(inst.tr, path);
			});
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_LIST') {
		inst.pop()
		.then(function(count) {
			return dirUtil.popTuples(inst, count)
			.then(function(path) {
				return directory.list(inst.tr, path);
			})
			.then(function(children) {
				inst.push(fdb.tuple.pack(children));
			});
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_EXISTS') {
		var path;
		inst.pop()
		.then(function(count) {
			return dirUtil.popTuples(inst, count)
			.then(function(p) {
				path = p;
				return directory.exists(inst.tr, path);
			})
			.then(function(exists) {
				logOp(util.format('exists (%s): %d', directory._path + (path ? path : ''), exists ? 1 : 0));
				inst.push(exists ? 1 : 0);
			});
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_PACK_KEY') {
		dirUtil.popTuples(inst)
		.then(function(keyTuple) {
			inst.push(directory.pack(keyTuple));
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_UNPACK_KEY') {
		inst.pop()
		.then(function(key) {
			logOp(util.format('unpack %s in subspace with prefix %s', fdb.buffer.printable(key), fdb.buffer.printable(directory.rawPrefix)));
			var tup = directory.unpack(key);
			for(var i = 0; i < tup.length; ++i)
				inst.push(tup[i]);
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_RANGE') {
		dirUtil.popTuples(inst)
		.then(function(tup) {
			var rng = directory.range(tup);
			inst.push(rng.begin);
			inst.push(rng.end);
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_CONTAINS') {
		inst.pop()
		.then(function(key) {
			inst.push(directory.contains(key) ? 1 : 0);
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_OPEN_SUBSPACE') {
		dirUtil.popTuples(inst)
		.then(function(path) {
			logOp(util.format('open_subspace (%s)', path));
			appendDir(directory.subspace(path));
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_LOG_SUBSPACE') {
		inst.pop()
		.then(function(prefix) {
			inst.tr.set(Buffer.concat([prefix, fdb.tuple.pack([self.dirIndex])]), directory.key());
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_LOG_DIRECTORY') {
		inst.pop()
		.then(function(prefix) {
			var exists;
			return directory.exists(inst.tr)
			.then(function(e) {
				exists = e;
				if(exists)
					return directory.list(inst.tr);
				else
					return [];
			})
			.then(function(children) {
				var logSubspace = new fdb.Subspace([self.dirIndex], prefix);
				inst.tr.set(logSubspace.get('path'), fdb.tuple.pack(directory.getPath()));
				inst.tr.set(logSubspace.get('layer'), fdb.tuple.pack([directory.getLayer()]));
				inst.tr.set(logSubspace.get('exists'), fdb.tuple.pack([exists ? 1 : 0]));
				inst.tr.set(logSubspace.get('children'), fdb.tuple.pack(children));
			});
		})(promiseCb);
	}
	else if(inst.op === 'DIRECTORY_STRIP_PREFIX') {
		inst.pop()
		.then(function(str) {
			if(!fdbUtil.buffersEqual(fdb.buffer(str).slice(0, directory.key().length), directory.key()))
				throw new Error('String ' + str + ' does not start with raw prefix ' + directory.key());

			inst.push(str.slice(directory.key().length));
		})(promiseCb);
	}
	else {
		throw new Error('Unknown op: ' + inst.op);
	}
};

module.exports = DirectoryExtension;

File: bindings/java/src/test/com/apple/foundationdb/test/DirectoryExtension.java
/*
 * DirectoryExtension.java
 *
 * This source file is part of the FoundationDB open source project
 *
 * Copyright 2013-2024 Apple Inc. and the FoundationDB project authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.apple.foundationdb.test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import java.util.concurrent.CompletableFuture;

import com.apple.foundationdb.Range;
import com.apple.foundationdb.directory.Directory;
import com.apple.foundationdb.directory.DirectoryLayer;
import com.apple.foundationdb.directory.DirectorySubspace;
import com.apple.foundationdb.subspace.Subspace;
import com.apple.foundationdb.tuple.Tuple;

class DirectoryExtension {
	List<Object> dirList = new ArrayList<Object>();
	int dirIndex = 0;
	int errorIndex = 0;

	DirectoryExtension() {
		dirList.add(DirectoryLayer.getDefault());
	}

	Directory directory() {
		return (Directory)dirList.get(dirIndex);
	}

	Subspace subspace() {
		return (Subspace)dirList.get(dirIndex);
	}

	void processInstruction(final Instruction inst) {
		try {
			//System.out.println("Processing operation " + inst.op + " (" + inst.context.instructionIndex + ")");
			DirectoryOperation op = DirectoryOperation.valueOf(inst.op);
			if(op == DirectoryOperation.DIRECTORY_CREATE_SUBSPACE) {
				Tuple prefix = DirectoryUtil.popTuple(inst).get();
				byte[] rawPrefix = (byte[])inst.popParam().get();
				dirList.add(new Subspace(prefix, rawPrefix));
			}
			else if(op == DirectoryOperation.DIRECTORY_CREATE_LAYER) {
				List<Object> params = inst.popParams(3).get();
				Subspace nodeSubspace = (Subspace)dirList.get(StackUtils.getInt(params.get(0)));
				Subspace contentSubspace = (Subspace)dirList.get(StackUtils.getInt(params.get(1)));
				boolean allowManualPrefixes = StackUtils.getInt(params.get(2)) == 1;

				if(nodeSubspace == null || contentSubspace == null)
					dirList.add(null);
				else
					dirList.add(new DirectoryLayer(nodeSubspace, contentSubspace, allowManualPrefixes));
			}
			else if(op == DirectoryOperation.DIRECTORY_CHANGE) {
				dirIndex = StackUtils.getInt(inst.popParam().get());
				if(dirList.get(dirIndex) == null)
					dirIndex = errorIndex;
			}
			else if(op == DirectoryOperation.DIRECTORY_SET_ERROR_INDEX) {
				errorIndex = StackUtils.getInt(inst.popParam().get());
			}
			else if(op == DirectoryOperation.DIRECTORY_CREATE_OR_OPEN || op == DirectoryOperation.DIRECTORY_OPEN) {
				List<String> path = DirectoryUtil.popPath(inst).get();
				byte[] layer = (byte[])inst.popParam().get();

				CompletableFuture<DirectorySubspace> dir;
				if(layer == null) {
					if(op == DirectoryOperation.DIRECTORY_CREATE_OR_OPEN)
						dir = directory().createOrOpen(inst.tcx, path);
					else
						dir = directory().open(inst.readTcx, path);
				}
				else {
					if(op == DirectoryOperation.DIRECTORY_CREATE_OR_OPEN)
						dir = directory().createOrOpen(inst.tcx, path, layer);
					else
						dir = directory().open(inst.readTcx, path, layer);
				}

				dirList.add(dir.get());
			}
			else if(op == DirectoryOperation.DIRECTORY_CREATE) {
				List<String> path = DirectoryUtil.popPath(inst).get();
				List<Object> params = inst.popParams(2).get();

				byte[] layer = (byte[])params.get(0);
				byte[] prefix = (byte[])params.get(1);

				CompletableFuture<DirectorySubspace> dir;
				if(layer == null && prefix == null)
					dir = directory().create(inst.tcx, path);
				else if(prefix == null)
					dir = directory().create(inst.tcx, path, layer);
				else {
					if(layer == null)
						layer = new byte[0];

					dir = directory().create(inst.tcx, path, layer, prefix);
				}

				dirList.add(dir.get());
			}
			else if(op == DirectoryOperation.DIRECTORY_MOVE) {
				List<List<String>> paths = DirectoryUtil.popPaths(inst, 2).get();
				DirectorySubspace dir = directory().move(inst.tcx, paths.get(0), paths.get(1)).get();
				dirList.add(dir);
			}
			else if(op == DirectoryOperation.DIRECTORY_MOVE_TO) {
				List<String> newAbsolutePath = DirectoryUtil.popPath(inst).get();
				DirectorySubspace dir = directory().moveTo(inst.tcx, newAbsolutePath).get();
				dirList.add(dir);
			}
			else if(op == DirectoryOperation.DIRECTORY_REMOVE) {
				int count = StackUtils.getInt(inst.popParam().get());
				List<List<String>> path = DirectoryUtil.popPaths(inst, count).get();
				if(path.size() == 0)
					directory().remove(inst.tcx).get();
				else
					directory().remove(inst.tcx, path.get(0)).get();
			}
			else if(op == DirectoryOperation.DIRECTORY_REMOVE_IF_EXISTS) {
				int count = StackUtils.getInt(inst.popParam().get());
				List<List<String>> path = DirectoryUtil.popPaths(inst, count).get();
				if(path.size() == 0)
					directory().removeIfExists(inst.tcx).get();
				else
					directory().removeIfExists(inst.tcx, path.get(0)).get();
			}
			else if(op == DirectoryOperation.DIRECTORY_LIST) {
				int count = StackUtils.getInt(inst.popParam().get());
				List<List<String>> path = DirectoryUtil.popPaths(inst, count).get();
				List<String> children;
				if(path.size() == 0)
					children = directory().list(inst.readTcx).get();
				else
					children = directory().list(inst.readTcx, path.get(0)).get();

				inst.push(Tuple.fromItems(children).pack());
			}
			else if(op == DirectoryOperation.DIRECTORY_EXISTS) {
				int count = StackUtils.getInt(inst.popParam().get());
				List<List<String>> path = DirectoryUtil.popPaths(inst, count).get();
				boolean exists;

				// In Java, DirectoryLayer.exists can return true without doing any reads.
				// Other bindings will always do a read, so we get a read version now to be compatible with that behavior.
				inst.readTcx.read(tr -> tr.getReadVersion().join());

				if(path.size() == 0)
					exists = directory().exists(inst.readTcx).get();
				else
					exists = directory().exists(inst.readTcx, path.get(0)).get();

				inst.push(exists ? 1 : 0);
			}
			else if(op == DirectoryOperation.DIRECTORY_PACK_KEY) {
				Tuple keyTuple = DirectoryUtil.popTuple(inst).get();
				inst.push(subspace().pack(keyTuple));
			}
			else if(op == DirectoryOperation.DIRECTORY_UNPACK_KEY) {
				byte[] key = (byte[])inst.popParam().get();
				Tuple tup = subspace().unpack(key);
				for(Object o : tup)
					inst.push(o);
			}
			else if(op == DirectoryOperation.DIRECTORY_RANGE) {
				Tuple tup = DirectoryUtil.popTuple(inst).get();
				Range range = subspace().range(tup);
				inst.push(range.begin);
				inst.push(range.end);
			}
			else if(op == DirectoryOperation.DIRECTORY_CONTAINS) {
				byte[] key = (byte[])inst.popParam().get();
				inst.push(subspace().contains(key) ? 1 : 0);
			}
			else if(op == DirectoryOperation.DIRECTORY_OPEN_SUBSPACE) {
				Tuple prefix = DirectoryUtil.popTuple(inst).get();
				dirList.add(subspace().subspace(prefix));
			}
			else if(op == DirectoryOperation.DIRECTORY_LOG_SUBSPACE) {
				final byte[] prefix = (byte[])inst.popParam().get();
				inst.tr.set(Tuple.from(dirIndex).pack(prefix), subspace().getKey());
			}
			else if(op == DirectoryOperation.DIRECTORY_LOG_DIRECTORY) {
				final byte[] prefix = (byte[])inst.popParam().get();
				boolean exists = directory().exists(inst.tr).get();
				List<String> children = exists ? directory().list(inst.tr).get() : new ArrayList<String>();
				Subspace logSubspace = new Subspace(new Tuple().add(dirIndex), prefix);
				inst.tr.set(logSubspace.pack("path"), Tuple.fromItems(directory().getPath()).pack());
				inst.tr.set(logSubspace.pack("layer"), new Tuple().add(directory().getLayer()).pack());
				inst.tr.set(logSubspace.pack("exists"), new Tuple().add(exists ? 1 : 0).pack());
				inst.tr.set(logSubspace.pack("children"), Tuple.fromItems(children).pack());
			}
			else if(op == DirectoryOperation.DIRECTORY_STRIP_PREFIX) {
				byte[] str = (byte[])inst.popParam().get();
				byte[] rawPrefix = subspace().getKey();

				if(str.length < rawPrefix.length)
					throw new RuntimeException("String does not start with raw prefix");

				for(int i = 0; i < rawPrefix.length; ++i)
					if(str[i] != rawPrefix[i])
						throw new RuntimeException("String does not start with raw prefix");

				inst.push(Arrays.copyOfRange(str, rawPrefix.length, str.length));
			}
			else {
				throw new RuntimeException("Unknown operation:" + inst.op);
			}
		}
		catch(Throwable t) {
			DirectoryUtil.pushError(inst, t, dirList);
		}
	}
}
